heartbeat/monitors/active/tcp/tls_test.go

```diff
@@ -0,0 +1,162 @@
+)
+
+// Tests that we can check a TLS connection with a cert for a SAN IP
+func TestTLSSANIPConnection(t *testing.T) {
+	ip, port, cert, certFile, teardown := setupTLSTestServer(t)
+	defer teardown()
+
+	event := testTLSTCPCheck(t, ip, port, certFile.Name(), monitors.StdResolver)
+	testslike.Test(
+		t,
+		lookslike.Strict(lookslike.Compose(
+			hbtest.TLSChecks(0, 0, cert),
+			hbtest.RespondingTCPChecks(),
+			hbtest.BaseChecks(ip, "up", "tcp"),
+			hbtest.SummaryChecks(1, 0),
+			hbtest.SimpleURLChecks(t, "ssl", ip, port),
+		)),
+		event.Fields,
+	)
+}
+
+func TestTLSHostname(t *testing.T) {
+	ip, port, cert, certFile, teardown := setupTLSTestServer(t)
+	defer teardown()
+
+	hostname := cert.DNSNames[0] // Should be example.com
+	resolver := monitors.CreateStaticResolver()
+	resolver.Add(hostname, ip)
+	event := testTLSTCPCheck(t, hostname, port, certFile.Name(), resolver)
+	testslike.Test(
+		t,
+		lookslike.Strict(lookslike.Compose(
+			hbtest.TLSChecks(0, 0, cert),
+			hbtest.RespondingTCPChecks(),
+			hbtest.BaseChecks(ip, "up", "tcp"),
+			hbtest.SummaryChecks(1, 0),
+			hbtest.SimpleURLChecks(t, "ssl", hostname, port),
+			hbtest.ResolveChecks(ip),
+		)),
+		event.Fields,
+	)
+}
+
+func TestTLSInvalidCert(t *testing.T) {
+	ip, port, cert, certFile, teardown := setupTLSTestServer(t)
+	defer teardown()
+
+	mismatchedHostname := "notadomain.elastic.co"
+	resolver := monitors.CreateStaticResolver()
+	resolver.Add(cert.DNSNames[0], ip)
+	resolver.Add(mismatchedHostname, ip)
+	event := testTLSTCPCheck(t, mismatchedHostname, port, certFile.Name(), resolver)
+
+	testslike.Test(
+		t,
+		lookslike.Strict(lookslike.Compose(
+			hbtest.RespondingTCPChecks(),
+			hbtest.BaseChecks(ip, "down", "tcp"),
+			hbtest.SummaryChecks(0, 1),
+			hbtest.SimpleURLChecks(t, "ssl", mismatchedHostname, port),
+			hbtest.ResolveChecks(ip),
+			lookslike.MustCompile(map[string]interface{}{
+				"error": map[string]interface{}{
+					"message": x509.HostnameError{Certificate: cert, Host: mismatchedHostname}.Error(),
+					"type":    "io",
+				},
+			}),
+		)),
+		event.Fields,
+	)
+}
+
+func setupTLSTestServer(t *testing.T) (ip string, port uint16, cert *x509.Certificate, certFile *os.File, teardown func()) {
+	// Start up a TLS Server
+	server, port := setupServer(t, httptest.NewTLSServer)
+
+	// Parse its URL
+	serverURL, err := url.Parse(server.URL)
+	require.NoError(t, err)
+
+	// Determine the IP address the server's hostname resolves to
+	ips, err := net.LookupHost(serverURL.Hostname())
+	require.NoError(t, err)
+	require.Len(t, ips, 1)
+	ip = ips[0]
+
+	// Parse the cert so we can test against it
+	cert, err = x509.ParseCertificate(server.TLS.Certificates[0].Certificate[0])
+	require.NoError(t, err)
+
+	// Save the server's cert to a file so heartbeat can use it
+	certFile = hbtest.CertToTempFile(t, cert)
+	require.NoError(t, certFile.Close())
+
+	return ip, port, cert, certFile, func() {
+		os.Remove(certFile.Name())
+		server.Close()
```

Comment:
defer?

Reply:
This function is returned in this way for the caller to call it in its own `defer`. The purpose of this function is to return a working TCP server, so we don't want to close the socket before the consumer has had a chance to use it :)

Reply:
I think you got me wrong. You can define `defer` in the body of the closure (`func()`). There is nothing preventing from calling defer in defer.

Reply:
@mtojek I may be missing something here, but why do we need a defer in the close? We don't have any branching that would prevent both statements from being executed.

Reply:
If you know that the server should be closed as the last (one of the few last) actions in this function, you can put it in defer.
It will prevent from any potential programming errors, when somebody adds a condition causing a premature exit from the function. 
In this case the `os.Remove` function returns an error if the operation fails. If anyone checks this error and decides to fail fast (`if err != nil { return err }`), then they might forget about closing the server.

(This is only suggestion, I understand that it may bring too little value at the moment.)

Reply:
I see what you mean. I'm not checking the error there because as you mention it's not really a huge value (it's a tempfile created by a test). That said, I think it does make sense to fail the test if the tempfile can't be removed (we have the test context here) and use defer to close the server.

I'll make that change and add the defer.

Reply:
Changed added!